libs/wallet/src/web3-react/connectors/Injected/WidgetEthereumProvider.ts

```diff
@@ -0,0 +1,384 @@
+/**
+ * ===========================================================================
+ * This is a modified version of the original iframe-provider library.
+ *
+ * The original had some deprecated methods and events, and it was hard to debug for the format used in the messages.
+ * This modified version makes use of the Widget library typed subscriptions
+ *
+ * https://github.com/ethvault/iframe-provider/blob/master/src/index.ts
+ *
+ *  ===========================================================================
+ */
+
+import { ProviderMessage } from '@walletconnect/ethereum-provider/dist/types/types'
+import { ProviderConnectInfo, ProviderRpcError } from '@web3-react/types'
+import { EventEmitter } from 'eventemitter3'
+
+import {
+  JsonRpcErrorResponseMessage,
+  JsonRpcRequestMessage,
+  JsonRpcSucessfulResponseMessage,
+  ProviderOnEventPayload,
+  ProviderRpcResponsePayload,
+  WidgetMethodsEmit,
+  WidgetMethodsListen,
+  listenToMessageFromWindow,
+  postMessageToWindow,
+} from '@cowprotocol/widget-lib'
+
+// By default timeout is 60 seconds
+const DEFAULT_TIMEOUT_MILLISECONDS = 60000
+
+const JSON_RPC_VERSION = '2.0'
+
+/**
+ * Options for constructing the iframe ethereum provider.
+ */
+interface IFrameEthereumProviderOptions {
+  // The origin to communicate with. Default '*'
+  targetOrigin?: string
+  // How long to time out waiting for responses. Default 60 seconds.
+  timeoutMilliseconds?: number
+
+  // The event source. By default we use the window. This can be mocked for tests, or it can wrap
+  // a different interface, e.g. workers.
+  eventSource?: Window
+
+  // The event target. By default we use the window parent. This can be mocked for tests, or it can wrap
+  // a different interface, e.g. workers.
+  eventTarget?: Window
+}
+
+/**
+ * This is what we store in the state to keep track of pending promises.
+ */
+interface PromiseCompleter<TResult, TErrorData> {
```

Comment:
Nitpick reaaaaly minor: don't prefix types with `T`. It's already implicit.

Reply:
they were copied from the library. Renamed to T and D (as its more usually for generics)

Reply:
Ah ok. If this was copied over the definitely ignore.